lib/ansible/playbook/taggable.py

```diff
@@ -47,16 +49,27 @@ def _load_tags(self, attr, ds):
         else:
             raise AnsibleError('tags must be specified as a list', obj=ds)
 
+    def _get_all_taggable_objects(self) -> t.Generator[Taggable, None, None]:
+        obj = self
+        while obj is not None:
+            yield obj
+            obj = obj._parent
+
+        if (role := getattr(self, "_role", Sentinel)) is not Sentinel:
+            yield role  # type: ignore[misc]
+
+        yield self.get_play()
+
     def evaluate_tags(self, only_tags, skip_tags, all_vars):
-        """ this checks if the current item should be executed depending on tag options """
+        """Check if the current item should be executed depending on the specified tags.
 
+        NOTE this method is assumed to be called only on Task objects.
+        """
         if self.tags:
             templar = Templar(loader=self._loader, variables=all_vars)
-            obj = self
-            while obj is not None:
+            for obj in self._get_all_taggable_objects():
                 if (_tags := getattr(obj, "_tags", Sentinel)) is not Sentinel:
                     obj._tags = _flatten_tags(templar.template(_tags))
-                obj = obj._parent
```

Comment:
a 'simpler' fix, but with more far reaching consequences is adding roles and plays as parents, so we don't have triple inheritance